tpu/src/test/java/tpu/TpuVmIT.java

```diff
@@ -0,0 +1,184 @@
+  public static void setUp()
+      throws IOException, ExecutionException, InterruptedException {
+    requireEnvVar("GOOGLE_APPLICATION_CREDENTIALS");
+    requireEnvVar("GOOGLE_CLOUD_PROJECT");
+
+    // Cleanup existing stale resources.
+    cleanUpExistingTpu("test-tpu-" + javaVersion, PROJECT_ID, ZONE);
+  }
+
+  @AfterAll
+  public static void cleanup() throws Exception {
+    DeleteTpuVm.deleteTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME);
+    TimeUnit.MINUTES.sleep(5);
+
+    // Test that TPUs is deleted
+    Assertions.assertThrows(
+        NotFoundException.class,
+        () -> GetTpuVm.getTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME));
+  }
+
+  @Test
+  @Order(1)
+  public void testCreateTpuVm() throws IOException, ExecutionException, InterruptedException {
+    final PrintStream out = System.out;
+    ByteArrayOutputStream stdOut = new ByteArrayOutputStream();
+    System.setOut(new PrintStream(stdOut));
+    CreateTpuVm.createTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME, ACCELERATOR_TYPE, VERSION);
+    TimeUnit.MINUTES.sleep(3);
+
+    assertThat(stdOut.toString()).contains("TPU VM created: " + TPU_VM_PATH_NAME);
+    stdOut.close();
+    System.setOut(out);
+  }
+
+  @Test
+  @Order(2)
+  public void testGetTpuVm() throws IOException {
+    Node node = GetTpuVm.getTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME);
+
+    assertNotNull(node);
+    assertThat(node.getName()).isEqualTo(TPU_VM_PATH_NAME);
+  }
+
+  @Test
+  @Order(2)
+  public void testListTpuVm() throws IOException {
+    TpuClient.ListNodesPagedResponse nodesList = ListTpuVms.listTpuVms(PROJECT_ID, ZONE);
+
+    assertNotNull(nodesList);
+    for (Node node : nodesList.iterateAll()) {
+      Assert.assertTrue(node.getName().contains("test-tpu"));
+    }
+  }
+
+  @Test
+  @Order(2)
+  public void testStopTpuVm() throws IOException, ExecutionException, InterruptedException {
+    StopTpuVm.stopTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME);
+    Node node = GetTpuVm.getTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME);
+
+    assertThat(node.getState()).isEqualTo(STOPPED);
+  }
+
+  @Test
+  @Order(3)
+  public void testStartTpuVm() throws IOException, ExecutionException, InterruptedException {
+    StartTpuVm.startTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME);
+    Node node = GetTpuVm.getTpuVm(PROJECT_ID, ZONE, TPU_VM_NAME);
+
+    assertThat(node.getState()).isEqualTo(READY);
+  }
+
+  public static void cleanUpExistingTpu(String prefixToDelete, String projectId, String zone)
+      throws IOException, ExecutionException, InterruptedException {
+    try (TpuClient tpuClient = TpuClient.create()) {
+      String parent = String.format("projects/%s/locations/%s", projectId, zone);
+      for (Node node : tpuClient.listNodes(parent).iterateAll()) {
+        String creationTime = formatTimestamp(node.getCreateTime());
+        String name = node.getName().substring(node.getName().lastIndexOf("/") + 1);
+        if (containPrefixToDeleteAndZone(node, prefixToDelete, zone)
+            && isCreatedBeforeThresholdTime(creationTime)) {
```

Comment:
If the test case takes < 5 minutes, do we just not delete the VM we created? Can we guarantee there won't be a TPU VM left running after the test runs once?

Reply:
Thank you! Fixed time to 30 minutes. After implementing all samples for TPU we can set it to 24 hours for example.